packages/headless/src/features/search/search-actions.ts

```diff
@@ -310,45 +311,21 @@ const buildFetchMoreRequest = (
 };
 
 function getFacets(state: StateNeededByExecuteSearch) {
-  return [
-    ...getFacetsInSameOrderAsResponse(state),
-    ...getFacetsNotInResponse(state),
-  ];
+  return [...getFacetsInOrder(state), ...getRemainingUnorderedFacets(state)];
 }
 
-function getFacetsInSameOrderAsResponse(state: StateNeededByExecuteSearch) {
-  const requests: AnyFacetRequest[] = [];
-  if (!state.search) {
-    return requests;
-  }
-  const responseFacets = state.search.response.facets;
-
-  responseFacets.forEach((f) => {
-    const request = findFacetRequest(state, f.facetId);
-    request && requests.push(request);
-  });
-
-  return requests;
+function getFacetsInOrder(state: StateNeededByExecuteSearch) {
+  return state?.search?.facetOrder
+    ? sortFacets(getAllFacets(state), state.search.facetOrder)
+    : [];
 }
 
-function findFacetRequest(state: StateNeededByExecuteSearch, facetId: string) {
-  const sets = [
-    state.facetSet,
-    state.numericFacetSet,
-    state.dateFacetSet,
-    state.categoryFacetSet,
-  ];
-
-  const targetSet = sets.find((set) => set && set[facetId]);
-  return targetSet ? targetSet[facetId] : undefined;
-}
-
-function getFacetsNotInResponse(state: StateNeededByExecuteSearch) {
-  const excludedFacetIds = new Set<string>();
-  const responseFacets = state.search?.response.facets;
-  responseFacets?.forEach((f) => excludedFacetIds.add(f.facetId));
-
-  return getAllFacets(state).filter((f) => !excludedFacetIds.has(f.facetId));
+function getRemainingUnorderedFacets(state: StateNeededByExecuteSearch) {
+  return isNullOrUndefined(state.search?.facetOrder)
+    ? getAllFacets(state)
+    : getAllFacets(state).filter(
+        (f) => state.search!.facetOrder!.indexOf(f.facetId) === -1
```

Comment:
I think you could arrange for TypeScript to "know" that a specific section of the state (in this case `search`) is defined if you end up in this scenario.

Otherwise, the various `!` peppered around in this function are kind of meh.

Reply:
Good point, the issue here was that `filter` takes a callback as a parameter, and TypeScript (rightfully) won't assume that because something isn't `null | undefined` before the execution of a callback it won't be during its execution.

I'm simply gonna extract `facetOrder` into a variable.